lib/backend/etcdv3/inline_cert_key.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2019 Tigera, Inc. All rights reserved.
+
+// This code has been based on code from etcd repository
+// to provide support for inline certificates and keys for calicoctl.
+// Below are the github links for the files from which the code has been borrowed.
+
+// Copyright 2015 The etcd Authors
+//     https://github.com/etcd-io/etcd/blob/release-3.3/pkg/transport/listener.go
+
+// Copyright 2016 The etcd Authors
+//     https://github.com/etcd-io/etcd/blob/release-3.3/pkg/tlsutil/tlsutil.go
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdv3
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+)
+
+// This struct is used to store the inline cert, key and CA cert data
+type TlsInlineCertKey struct {
+	Cert   string
+	Key    string
+	CACert string
+}
+
+// ClientConfigInlineCertKey() returns a pointer to tls.Config struct object with certificate data
+// for client creation using only the inline certificate, key and CA certificate data.
+func (info TlsInlineCertKey) ClientConfigInlineCertKey() (*tls.Config, error) {
+	var cfg *tls.Config
+	var err error
+
+	if info.Cert != "" && info.Key != "" {
+		cfg, err = info.baseCertConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if info.CACert != "" {
+		cfg.RootCAs, err = newCertPool(info.CACert)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cfg, nil
+}
+
+// baseCertConfig() populates tls struct with certificate data
+func (info TlsInlineCertKey) baseCertConfig() (*tls.Config, error) {
+	if info.Key == "" || info.Cert == "" {
+		return nil, fmt.Errorf("Key and Cert must both be present[key: %v, cert: %v]", info.Key, info.Cert)
```

Comment:
This will log out the key data, which is sensitive.  Please remove the key from the log.

Reply:
Acknowledged and updated log statement.